src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private Object[] elements;
+    private int size;
 
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
+        this.size = 0;
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T element) {
+        if (size == elements.length) {
+            resizeArray();
+        }
+        elements[size++] = element;
+    }
 
+    @Override
+    public void add(T element, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Некоректний індекс: " + index);
+        }
+        if (size == elements.length) {
+            resizeArray();
+        }
+        System.arraycopy(
+                elements, index,
+                elements, index + 1,
+                size - index);
+        elements[index] = element;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T remove(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Некоректний індекс: " + index);
+        }
+        T removedElement = getElement(index);
+        System.arraycopy(
+                elements, index + 1,
+                elements, index,
+                size - index - 1);
+        elements[--size] = null;
+        return removedElement;
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public T remove(T element) {
+        for (int i = 0; i < size; i++) {
+            if (elements[i] == null) {
+                if (element == null) {
+                    return remove(i);
+                }
+            } else if (elements[i].equals(element)) {
+                return remove(i);
+            }
```

Comment:
combine conditions, there shouldn't be inner if-else. you have the same body in two cases